app/app.go

```diff
@@ -0,0 +1,168 @@
+package app
+
+import (
+	"context"
+	"time"
+
+	dockerClient "github.com/docker/docker/client"
+	"github.com/juju/errors"
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	calicov3 "github.com/projectcalico/libcalico-go/lib/clientv3"
+
+	"github.com/projecteru2/barrel/driver"
+	calicoDriver "github.com/projecteru2/barrel/driver/calico"
+	fixedipDriver "github.com/projecteru2/barrel/driver/fixedip"
+	"github.com/projecteru2/barrel/proxy"
+	dockerProxy "github.com/projecteru2/barrel/proxy/docker"
+	"github.com/projecteru2/barrel/store"
+	"github.com/projecteru2/barrel/store/etcd"
+	"github.com/projecteru2/barrel/utils"
+	"github.com/projecteru2/barrel/vessel"
+)
+
+// Application .
+type Application struct {
+	Mode                   string
+	DockerGID              int64
+	DockerDaemonUnixSocket string
+	Hosts                  []string
+	DialTimeout            time.Duration
+	CertFile               string
+	KeyFile                string
+	ShutdownTimeout        time.Duration
+}
+
+// Run .
+func (app Application) Run() error {
+	appRunner := appRunner{app: &app}
+	if app.Mode == "default" {
+		return appRunner.defaultMode()
+	} else if app.Mode == "proxy-only" {
+		return appRunner.proxyOnlyMode()
+	} else if app.Mode == "network-plugin-only" {
+		return appRunner.networkPluginOnlyMode()
+	} else {
+		return errors.New("Unrecognized barrel mode, support only [ --mode default | proxy-only | network-plugin-only ]")
+	}
+}
+
+type appRunner struct {
+	app          *Application
+	apiConfig    *apiconfig.CalicoAPIConfig
+	dockerClient *dockerClient.Client
+	vessel       vessel.Helper
+	etcdClient   store.Store
+	calicoClient calicov3.Interface
+}
+
+func (app *appRunner) proxy() service {
+	return func(ctx context.Context) error {
+		var (
+			serv proxy.Service
+			err  error
+		)
+		if serv, err = dockerProxy.NewProxyService(
+			dockerProxy.Config{},
+			app.vessel,
+		); err != nil {
+			return err
+		}
+		return serv.Serve(ctx)
+	}
+}
+
+func (app *appRunner) calicoNetworkPlugin() service {
+	calicoClient, err1 := app.getCalicoClient()
+	dockerClient, err2 := app.getDockerClient()
+	return func(ctx context.Context) error {
+		if err1 != nil {
+			return err1
+		}
+		if err2 != nil {
+			return err2
+		}
+		serv := driver.Service{
+			Ipam: calicoDriver.Ipam{
+				CalicoIPAllocator: app.vessel.CalicoIPAllocator(),
+				ObjectLogger:      utils.ObjectLogger{ObjectName: "calicoIpam"},
+			},
+			Driver: calicoDriver.NewDriver(
+				calicoClient,
+				dockerClient,
+			),
+		}
+		return serv.Serve(ctx)
+	}
+}
+
+func (app *appRunner) barrelNetworkPlugin() service {
+	calicoClient, err1 := app.getCalicoClient()
+	dockerClient, err2 := app.getDockerClient()
```

Comment:
im not sure why put them out of the below func.

Reply:
'cause the below func will execute asynchronously, there might be concurrency issues. planning a changing here, to allow a earlier failure